supervision/utils/file.py

```diff
@@ -1,10 +1,121 @@
+import csv
 import json
 from pathlib import Path
-from typing import List, Optional, Union
+from typing import Any, Dict, List, Optional, Union
 
 import numpy as np
 import yaml
 
+from supervision.detection.core import Detections
+
+
+class CSVSink:
+    """
+    A utility class for saving detection data to a CSV file. This class is designed to
+    efficiently serialize detection objects into a CSV format, allowing for the inclusion of
+    bounding box coordinates and additional attributes like confidence, class ID, and tracker ID.
+
+    The class supports the capability to include custom data alongside the detection fields,
+    providing flexibility for logging various types of information.
+
+    Args:
+        filename (str): The name of the CSV file where the detections will be stored.
+                        Defaults to 'output.csv'.
+
+    Usage:
+        ```python
+        from supervision.utils.detections import Detections
+        # Initialize CSVSink with a filename
+        csv_sink = CSVSink('my_detections.csv')
+
+        # Assuming detections is an instance of Detections containing detection data
+        detections = Detections(...)
+
+        # Open the CSVSink context, append detection data, and close the file automatically
+        with csv_sink as sink:
+            sink.append(detections, custom_data={'frame': 1})
+        ```
+    """
+
+    def __init__(self, filename: str = "output.csv"):
+        self.filename = filename
+        self.file: Optional[open] = None
+        self.writer: Optional[csv.writer] = None
+        self.header_written = False
+        self.fieldnames = []  # To keep track of header names
+
+    def __enter__(self) -> "CSVSink":
+        self.open()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[type],
+        exc_val: Optional[Exception],
+        exc_tb: Optional[Any],
+    ) -> None:
+        self.close()
+
+    def open(self) -> None:
+        self.file = open(self.filename, "w", newline="")
+        self.writer = csv.writer(self.file)
+
+    def close(self) -> None:
+        if self.file:
+            self.file.close()
+
+    def append(
+        self, detections: Detections, custom_data: Dict[str, Any] = None
+    ) -> None:
+        if not self.writer:
+            raise Exception(
+                f"Cannot append to CSV: The file '{self.filename}' is not open. Ensure that the 'open' method is called before appending data."
```

Comment:
Let's break the line here. Make it under 88 characters. Otherwise, Ruff will complain.